pkg/natdiscovery/natdiscovery.go

```diff
@@ -0,0 +1,188 @@
+	GetEndpointIP(endpointName string) (string, error)
+}
+
+type udpWriteFunction func(b []byte, addr *net.UDPAddr) (int, error)
+type findSrcIPFunction func(destinationIP string) (string, error)
+
+type natDiscovery struct {
+	sync.Mutex
+	localEndpoint   *types.SubmarinerEndpoint
+	remoteEndpoints map[string]*remoteEndpointNAT
+	requestCounter  uint64
+	serverUDPWrite  udpWriteFunction
+	findSrcIP       findSrcIPFunction
+	serverPort      int
+}
+
+func New(localEndpoint *types.SubmarinerEndpoint) (Interface, error) {
+	return newNatDiscovery(localEndpoint)
+}
+
+func newNatDiscovery(localEndpoint *types.SubmarinerEndpoint) (*natDiscovery, error) {
+	port, err := extractNATDiscoveryPort(localEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	requestCounter, err := randomRequestCounter()
+	if err != nil {
+		return nil, err
+	}
+
+	return &natDiscovery{
+		localEndpoint:   localEndpoint,
+		serverPort:      port,
+		remoteEndpoints: map[string]*remoteEndpointNAT{},
+		findSrcIP:       findPreferredSourceIP,
+		requestCounter:  requestCounter,
+	}, nil
+}
+
+func randomRequestCounter() (uint64, error) {
+	n, err := rand.Int(rand.Reader, big.NewInt(1000))
+	if err != nil {
+		return 0, errors.Wrapf(err, "generatimg random request counter")
+	}
+
+	return n.Uint64(), nil
+}
+
+var errorNoNatDiscoveryPort = errors.New("no nat discovery port in endpoint")
+
+func extractNATDiscoveryPort(endpoint *types.SubmarinerEndpoint) (int, error) {
+	if endpoint.Spec.PortConfig == nil {
+		return 0, errorNoNatDiscoveryPort
+	}
+
+	if portString, ok := endpoint.Spec.PortConfig[v1.NATDiscoveryPortConfig]; ok {
+		port, err := strconv.Atoi(portString)
+		if err != nil {
+			return 0, errors.Wrapf(err, "error trying to extract port number from %q in %s",
+				portString, v1.NATDiscoveryPortConfig)
+		}
+
+		return port, nil
+	}
+
+	return 0, errorNoNatDiscoveryPort
+}
+
+func (nd *natDiscovery) Run(stopCh <-chan struct{}, natResultCh chan remoteEndpointNAT) error {
+	err := nd.runListener(stopCh)
+	if err != nil {
+		return err
+	}
+
+	go wait.Until(func() {
+		nd.checkEndpointList()
+	}, time.Second, stopCh)
+
+	return nil
+}
+
+func (nd *natDiscovery) AddEndpoint(endpoint *types.SubmarinerEndpoint) {
+	remoteNAT := newRemoteEndpointNAT(endpoint)
+
+	// support a remote cluster endpoint which still hasn't implemented this protocol
+	if _, err := extractNATDiscoveryPort(endpoint); err == errorNoNatDiscoveryPort {
+		remoteNAT.useLegacyNATSettings()
+	}
+
+	nd.Lock()
+	defer nd.Unlock()
+
+	nd.remoteEndpoints[endpoint.Spec.CableName] = remoteNAT
+}
+
+func (nd *natDiscovery) RemoveEndpoint(endpointName string) {
+	nd.Lock()
+	defer nd.Unlock()
+	delete(nd.remoteEndpoints, endpointName)
+}
+
+func (nd *natDiscovery) checkEndpointList() {
+	nd.Lock()
+	defer nd.Unlock()
+
+	for _, endpointNAT := range nd.remoteEndpoints {
+		if endpointNAT.hasTimedOut() {
+			endpointNAT.useLegacyNATSettings()
+		} else if endpointNAT.shouldCheck() {
```

Comment:
Once it transitions to public IP or private IP via a valid response, it should stay in that state. However `hasTimedOut` will eventually return true once the `totalTimeout` has elapsed and thus we'll revert to `useLegacyNATSettings`. We should restructure this so that doesn't happen. I think the simplest is:

```
if endpointNAT.shouldCheck() {
    if endpointNAT.hasTimedOut() {
        endpointNAT.useLegacyNATSettings()
        continue
    }

    err := nd.sendCheckRequest(endpointNAT)
    ...
}
```

Another option is to modify `hasTimedOut` to check the state but `shouldCheck` already does that.  `hasTimedOut` may be slightly delayed by the `recheckTime` but I think that's OK.

Reply:
Opps, that's a good one.... thanks tom!


From publicIP to privateIP we allow a grace time where we can still transition to privateIP (since we test both private & public in parallel) you could have reachability to both, but we will always prefer the privateIP.
After that grace time we stop accepting responses from privateIP, because it could be a forged packet to try to cheat the gateway to switch IPs (and break connectivity)